drivers/sdhc/rcar_mmc.c

```diff
@@ -0,0 +1,1597 @@
+			cmd->response[2], cmd->response[3]);
+	} else {
+		cmd->response[0] = rcar_mmc_read_reg32(cfg, RCAR_MMC_RSP10);
+		LOG_DBG("Response %u\n\t[0]: 0x%08x", response_type, cmd->response[0]);
+	}
+}
+
+/**
+ * @brief Configure CMD register for tx/rx data
+ *
+ * @note in/out parameters should be checked by a caller function.
+ *
+ * @param cmd MMC command
+ * @param data MMC data buffer for tx/rx
+ *
+ * @retval partial configuration of CMD register
+ */
+static uint32_t rcar_mmc_gen_data_cmd(struct sdhc_command *cmd,
+	struct sdhc_data *data)
+{
+	uint32_t cmd_reg = RCAR_MMC_CMD_DATA;
+
+	switch (cmd->opcode) {
+	case MMC_SEND_EXT_CSD:
+	case SD_READ_SINGLE_BLOCK:
+		cmd_reg |= RCAR_MMC_CMD_RD;
+		break;
+	case SD_READ_MULTIPLE_BLOCK:
+		cmd_reg |= RCAR_MMC_CMD_RD;
+		cmd_reg |= RCAR_MMC_CMD_MULTI;
+		break;
+	case SD_WRITE_MULTIPLE_BLOCK:
+		cmd_reg |= RCAR_MMC_CMD_MULTI;
+		break;
+	case SD_WRITE_SINGLE_BLOCK:
+		/* fall through */
+	default:
+		break;
+	}
+
+	if (data->blocks > 1) {
+		cmd_reg |= RCAR_MMC_CMD_MULTI;
+	}
+
+	return cmd_reg;
+}
+
+/**
+ * @brief Transmit/Receive data to/from MMC
+ *
+ * Sends/Receives data to/from the MMC controller.
+ *
+ * @note in/out parameters should be checked by a caller function.
+ *
+ * @param dev MMC device
+ * @param cfg MMC driver configuration
+ * @param cmd MMC command
+ * @param data MMC data buffer for tx/rx
+ * @param is_read it is read or write operation
+ *
+ * @retval 0 tx/rx was successful
+ * @retval -EINVAL: invalid block size
+ * @retval -ETIMEDOUT: timed out while tx/rx
+ * @retval -EIO: I/O error
+ * @retval -EILSEQ: communication out of sync
+ */
+static int rcar_mmc_rx_tx_data(const struct device *dev,
+	const struct mmc_rcar_cfg *cfg,
+	struct sdhc_command *cmd,
+	struct sdhc_data *data,
+	bool is_read)
+{
+	struct mmc_rcar_data *dev_data = dev->data;
+	uint32_t info1_reg;
+	uint32_t block;
+	int ret = 0;
+	uint32_t info2_poll_flag = is_read ? RCAR_MMC_INFO2_BRE : RCAR_MMC_INFO2_BWE;
+
+	rcar_mmc_write_reg32(cfg, RCAR_MMC_INFO1_MASK,
+			(uint32_t)~RCAR_MMC_INFO1_CMP);
+
+	rcar_mmc_write_reg32(cfg, RCAR_MMC_INFO2_MASK,
+			~(info2_poll_flag | RCAR_MMC_INFO2_ERRORS));
+
+#ifdef CONFIG_RCAR_MMC_DMA_SUPPORT
+#error "Add support of DMA request"
+#else
+	if ((data->block_size % dev_data->width_access_sd_buf0) ||
+	    (data->block_size < dev_data->width_access_sd_buf0)) {
+		/* TODO: fix me */
+		LOG_ERR("%s: block size (%u) less or not align on SD BUF0 access width (%hhu)",
+			dev->name, data->block_size, dev_data->width_access_sd_buf0);
+		return -EINVAL;
+	}
+
+	for (block = 0; block < data->blocks; block++) {
+		uint64_t *buf = (uint64_t *)data->data;
+		uint32_t info2_reg;
+		uint32_t w_off; /* word offset in a block */
+		/* wait until the buffer is filled with data */
+		ret = rcar_mmc_poll_reg_flags_check_err(cfg, RCAR_MMC_INFO2,
+						info2_poll_flag, info2_poll_flag, true);
+		if (ret) {
+			return ret;
+		}
+
+		buf += (block * data->block_size) / sizeof(*buf);
```

Comment:
That's an interesting way to manage pointer to the buffer. 
Could you please move it to the declaration? 
```
		uint64_t *buf = (uint64_t *)data->data + (block * data->block_size) / sizeof(*buf);
```

Because it confuses me for the second time. What are you doing there is initializing `buf` value in two steps. And for reader it looks like a bug, because reader except to see the traditional approach, where `buf` is declared out of `for` scope and it is increased every iteration.

Reply:
> And for reader it looks like a bug,

as for me, it isn't, because I can memorize at least a few lines that I have seen before
ok, I'll change it for better understanding of others
thanks for comment!